core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -736,7 +737,7 @@ private boolean expandStar(List<SqlNode> selectItems, Set<String> aliases,
       if (!hasDynamicStruct || Bug.CALCITE_2400_FIXED) {
         // If some fields before star identifier,
         // we should move offset.
-        int offset = calculatePermuteOffset(selectItems);
+        int offset = Math.min(calculatePermuteOffset(selectItems), originalSize);
```

Comment:
I see this is gated on CALCITE_2400.
What is the relationship between this bug and CALCITE-2400?

Reply:
@mihaibudiu  It was possible to get an `IndexOutOfBoundsException` when correcting the field order (for CALCITE-2400). We need to make sure that the offset is not beyond the `selectItems` list.

Reply:
My question was whether the `|| Bug.CALCITE_2400_FIXED` is still needed after your fix.

Reply:
I don't see any use for the `Bug.CALCITE_2400_FIXED` flag anymore. Removed it.